contrib/sacrebleu/sacrebleu.py

```diff
@@ -0,0 +1,894 @@
+    :return: the set of processed files
+    """
+
+    # if not data.has_key(test_set):
+    #     return None
+
+    dataset = data[test_set]['data']
+    outdir = os.path.join(SACREBLEU, test_set)
+    if not os.path.exists(outdir):
+        logging.info('Creating {}'.format(outdir))
+        os.makedirs(outdir)
+
+    tarball = os.path.join(outdir, os.path.basename(dataset))
+    rawdir = os.path.join(outdir, 'raw')
+    if not os.path.exists(tarball):
+        # TODO: check MD5sum
+        logging.info("Downloading {} to {}".format(dataset, tarball))
+        with urllib.request.urlopen(dataset) as f, open(tarball, 'wb') as out:
+            out.write(f.read())
+
+        # Extract the tarball
+        logging.info('Extracting {}'.format(tarball))
+        tar = tarfile.open(tarball)
+        tar.extractall(path=rawdir)
+
+    found = []
+
+    # Process the files into plain text
+    languages = data[test_set].keys() if langpair is None else [langpair]
+    for pair in languages:
+        if '-' not in pair:
+            continue
+        src, tgt = pair.split('-')
+        rawfile = os.path.join(rawdir, data[test_set][pair][0])
+        outfile = os.path.join(outdir, '{}.{}'.format(pair, src))
+        process_to_text(rawfile, outfile)
+        found.append(outfile)
+
+        rawfile = os.path.join(rawdir, data[test_set][pair][1])
+        outfile = os.path.join(outdir, '{}.{}'.format(pair, tgt))
+        process_to_text(rawfile, outfile)
+        found.append(outfile)
+
+    return found
+
+
+BLEU = namedtuple('BLEU', 'score, counts, totals, precisions, bp, sys_len, ref_len')
+
+
+def compute_bleu(correct: List[int], total: List[int], sys_len: int, ref_len: int, smooth = 'none', smooth_floor = 0.01):
+    """Computes BLEU score from its sufficient statistics. Adds smoothing.
+
+    :param correct: List of counts of correct ngrams, 1 <= n <= NGRAM_ORDER
+    :param total: List of counts of total ngrams, 1 <= n <= NGRAM_ORDER
+    :param sys_len: The cumulative system length
+    :param ref_len: The cumulative reference length
+    :param smooth: The smoothing method to use
+    :param smooth_floor: The smoothing value added, if smooth method 'floor' is used
+    """
+
+    precisions = [0 for x in range(NGRAM_ORDER)]
+
+    smooth_mteval = 1
+    for n in range(NGRAM_ORDER):
+        if total[n] == 0:
+            precisions[n] = 0
+        elif correct[n] == 0:
+            if smooth == 'exp':
+                smooth_mteval *= 2
+                precisions[n] = 100. / (smooth_mteval * total[n])
+            elif smooth == 'floor':
+                precisions[n] = 100. * smooth_floor / total[n]
+        else:
+            precisions[n] = 100. * correct[n] / total[n]
+
+    brevity_penalty = 1.0
+    if sys_len < ref_len:
+        brevity_penalty = math.exp(1 - ref_len / sys_len) if sys_len > 0 else 0.0
+
+    bleu = 1. * brevity_penalty * math.exp(sum(map(my_log, precisions)) / NGRAM_ORDER)
```

Comment:
If none of the hypothesis had any 4-grams, do we still want to divide by 4? (cf. effective_order in bleu.py)

If all hypothesis are completely empty (not just missing some high-order n-grams), looks like the BLEU will not be 0 - is it expected?

BTW, what's the point of multiplying by 1?

Reply:
Yes, this is correct. If there are 0 four-grams and there was no smoothing, it won't matter. This also passes all sockeye tests and is exactly the same as metal-v13a.pl in Moses.

1-multiplication was a holdover, I'll remove it.